shutter/shutter.md

```diff
@@ -0,0 +1,284 @@
+## Encrypting RPC Server
+
+The Encrypting RPC Server exposes the standard Ethereum RPC API. Whenever a user submits a transaction via `eth_sendRawTransaction`, the server will attempt to encrypt it. To this end, it keeps track of the eon keys from the Eon Key Broadcast Contract as well as the set of participating validators from the Validator Registry. It selects a slot in the near future whose proposer is registered and encrypts the transaction for this slot. It then submits it to the Sequencer Contract.
+
+### Interfaces with
+
+- Key Broadcast Contract: To receive eon keys
+- Sequencer Contract: To provide encrypted transactions to
+- Validator Registry: To check registered validators
+
+### Interface
+
+See [https://ethereum.github.io/execution-apis/api-documentation/](https://ethereum.github.io/execution-apis/api-documentation/).
+
+## Decryption Key Relay
+
+The Decryption Key Relay allows users to listen to decryption keys without having to join the libp2p gossip network themselves. It subscribes to both the `decryptionKey` and `decryptionKeyShare` topics. It applies the filtering logic described in the Keyper-section. It relays the remaining messages to the clients that suscribed to the interface described below.
+
+### Interfaces with
+
+- Keypers: To receive keys from
+- Validators: To send keys to
+
+### Interface
+
+The relay provides an HTTP GET endpoint at path `/v1/decryption_keys` with the following required query parameters:
+
+- `instance_id`: A decimal encoded instance ID
+
+If the relay is connected to a gossip network with different instance ID, it responds with status code `404`. Otherwise, an SSE stream is opened that yields items of the following format:
+
+```jsx
+event: decryption_key;
+data: data;
+```
+
+`data` is the `0x`-prefixed, hex encoded decryption key protocol buffer defined under “Keypers”.
+
+As the relay is meant to be an only provisional solution it does not perform any authentication.
+
+## Encodings
+
+This section describes how various data types used above are canonically encoded.
+
+### P2P Messages
+
+Gossiped p2p messages are encoded as one of the following protocol buffers:
+
+```jsx
+message DecryptionKeyShare {
+    uint64 instanceID = 1;
+    uint64 eon = 2;
+    bytes epochID = 3;
+    uint64 keyperIndex = 4;
+    bytes share = 5;
+}
+
+message DecryptionKey {
+    uint64 instanceID = 1;
+    uint64 eon = 2;
+    bytes epochID = 3;
+    bytes key = 4;
+}
+```
+
+- `instanceID`: The ID of the instance for which the message is relevant.
+- `eon`: An eon index.
+- `epochID`: A big-endian encoded slot number. Note that this field is unrelated to the epochs of the beacon chain.
+- `keyperIndex`: A zero-based number indexing a keyper set.
+- `key`: A decryption key encoded as described below.
+- `share`: A decryption key share encoded as described below.
+
+Messages are wrapped in the following envelope:
+
+```jsx
+message Envelope {
+    uint32 version = 1 ;
+    google.protobuf.Any message = 2;
+}
```

Comment:
Following beacon chain's existing design I think it's better to have versioning on the topic instead of the message. If we want to do versioning on the message let's just add the version field the only two pb `DecryptionKeyShare` and `DecryptionKey`